utils_cv/detection/dataset.py

```diff
@@ -96,38 +150,64 @@ def __init__(
             self.train_ds,
             batch_size=self.batch_size,
             shuffle=True,
-            num_workers=4,
+            num_workers=db_num_workers(),
             collate_fn=collate_fn,
         )
 
         self.test_dl = DataLoader(
             self.test_ds,
             batch_size=self.batch_size,
             shuffle=False,
-            num_workers=4,
+            num_workers=db_num_workers(),
             collate_fn=collate_fn,
         )
 
-    def _get_labels(self) -> List[str]:
+    def _read_annos(self) -> List[str]:
         """ Parses all Pascal VOC formatted annotation files to extract all
         possible labels. """
-        # TODO implement:
-        # - option to bypass this setup set if users already know label names
-        # - option to do 'active' loading so that we preload all annotations
-        # during setup time (instead of lazy loading) - this will let us
-        # do things like get average box size of all the data, etc..
 
+        # If im_dir specified then get image paths and respective annotations
+        # from scanning that directory. Otherwise find all annotation files
+        # and get path to image from annotation.
+        if self.im_dir is None:
+            anno_filenames = sorted(os.listdir(self.root / self.anno_dir))
+            self.im_paths = []
+        else:
+            im_filenames = sorted(os.listdir(self.root / self.im_dir))
+            anno_filenames = [
+                os.path.splitext(s)[0] + ".xml" for s in im_filenames
+            ]
+            self.im_paths = [
+                os.path.join(self.root / self.im_dir, s) for s in im_filenames
+            ]
+
+        # Parse all annotations
+        self.anno_paths = []
+        self.anno_bboxes = []
+        for anno_filename in anno_filenames:
+            anno_path = self.root / self.anno_dir / str(anno_filename)
```

Comment:
maybe we should check here that the anno_path is valid since we are assuming that images and annotation files have the same name. Which is fair, but just in case.
`assert is_valid_file(anno_path)`